### DISCUSSION TITLE
Workflow Optimization and Collaboration

### DISCUSSION THREAD

**Discussion by codetricity** [camera360developers/theta-client-guide#3] [General]
## pull requests, merges, commits, issues, discussions
  Ideas, suggestions and observations on using GitHub and git effectively for collaboration. 
  * Put links to best practices. Avoid personal opinion on workflow optimization unless it is based on experience.
  * Brainstorming on best practice is acceptable. However, it's best to research first, then put an opinion based on research into the discussion
  * off-topic comments about fun activities are acceptable. We can socialize.
  * humor and jokes are OK

**Comment by codetricity** [comment]
Based on discussions on March 7, 2023, @AirTechWick (Erik) to use the single branch below until he completes the Windows developer workflow guide for demo-react-native.

https://github.com/theta360developers/theta-client-guide/tree/erik-react-native-windows

* as much as possible only @AirTechWick should edit the file [react-native-install.md](https://github.com/theta360developers/theta-client-guide/blob/main/docs/react-native-install.md)
* Erik should make 1-3 pull requests per day.  This will help us to practice merging his work

**Comment by codetricity** [comment]
I just added a commit with the prefix `refactor: `.

I saw the syntax in the article below.

https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/

The official docs for the conventional commit specification is here:

https://www.conventionalcommits.org/en/v1.0.0/

We've been using `feat: ` and `fix: `.  I've started to use `refactor: ` as well.

**Comment by AirTechWick** [comment]
[Git Workflows](https://www.atlassian.com/git/tutorials/comparing-workflows)

Found this article on Git/Github workflows. I believe the "Centralized Workflow" is the one we are practicing currently, and works with people you give access to make pull requests to the main repository. 

The article mentions:
"The Forking Workflow is most often seen in public open source projects.

The main advantage of the Forking Workflow is that contributions can be integrated without the need for everybody to push to a single central repository. Developers push to their own server-side repositories, and only the project maintainer can push to the official repository." 

Basically, If I fork anyone's repository, I can make changes in my own fork and make a pull request to merge my fork of the repository.

**Reply by codetricity** [reply]
Thank you for looking into this article.

IMO, the "Centralized Workflow" is easier.  We do not have an "open source project" that requires a fork for our workflow.  We may fork the theta-client project from RICOH and make pull requests in the future.  Like this:

  https://github.com/ricohapi/theta-client/pull/10

For the "Centralized Workflow", I do not use rebase and actually do not know how to use it or when to use it.

---

## Mary rebases on top of John’s commit(s)

![image](https://user-images.githubusercontent.com/5799860/223801836-38d24cb0-1a55-4f6a-945d-ce5efdb2990d.png)


Mary can use git pull to incorporate upstream changes into her repository. This command is sort of like svn update—it pulls the entire upstream commit history into Mary’s local repository and tries to integrate it with her local commits:

 ```
git pull --rebase origin main
```

The `--rebase` option tells Git to move all of Mary’s commits to the tip of the main branch after synchronising it with the changes from the central repository, as shown below:

![image](https://user-images.githubusercontent.com/5799860/223802118-0f0d99c3-d0bf-4e22-a783-757eb0f7f5cc.png)


The pull would still work if you forgot this option, but you would wind up with a superfluous “merge commit” every time someone needed to synchronize with the central repository. For this workflow, it’s always better to rebase instead of generating a merge commit.

---

It seems like we should try using `--rebase` after studying the article that Erik shared.

## my first rebase

```
git pull --rebase origin main
From github.com:theta360developers/theta-client-guide
 * branch            main       -> FETCH_HEAD
Updating 21d19e5..daf62e8
Fast-forward
 mkdocs.yml | 2 ++
 1 file changed, 2 insertions(+)
```

**Comment by codetricity** [comment]
# remove stale branches

## problem

![image](https://user-images.githubusercontent.com/5799860/223883232-b538d7fd-ba88-40f5-a982-c1c0bb30c831.png)

## solution

```
git branch -d -r origin/erik-react-native
Deleted remote-tracking branch origin/erik-react-native (was 0ff3bdf).
```

## verify

```
git branch -a
  craig-camera-comparison
  gh-pages
* main
  remotes/origin/HEAD -> origin/main
  remotes/origin/craig-flutter
  remotes/origin/erik-react-native-windows
  remotes/origin/gh-pages
  remotes/origin/main
```

**Comment by jcasman** [comment]
@AirTechWick Did you remove stale branches? I did a `git pull` and ran `git branch -a` and it lists remotes/origin/erik-react-native . Here's my full output:

```
MacBook-Air-5:theta-client-guide jcasman$ git branch -a
* main
  remotes/origin/HEAD -> origin/main
  remotes/origin/craig-camera-compare
  remotes/origin/craig-flutter
  remotes/origin/erik-react-native
  remotes/origin/erik-react-native-windows
  remotes/origin/gh-pages
  remotes/origin/main
```

**Reply by codetricity** [reply]
I think that `erik-react-native` is old and no longer used.

**Reply by AirTechWick** [reply]
I think today I was on the old branch somehow, so I think the workflow is : right after I'm done using a branch to remove the stale branch on my own computer

**Comment by codetricity** [comment]
no problems with the last pull request from Erik.

<img width="591" alt="image" src="https://user-images.githubusercontent.com/5799860/224833190-5d190725-21b0-4b5e-92c3-c35d8b1b9a3f.png">

The process of merging into the main branch and deploying was easy.

An ambitious and potentially problematic approach is for  Jesse to edit the spelling and grammar directly on his own branch and make a pull request.

Erik would then need to pull down the changes tomorrow morning.

**Comment by jcasman** [comment]
I have made edits using a new branch jcasman-theta-client-guide. I am having trouble running git push -u origin jcasman-theta-client-guide due to my ssh keys not being set up correctly. I should have edits uploaded soon.

**Reply by codetricity** [reply]
I don't see the pull request as of 6am PT on March 14.

Did you read the GitHub documentation on how to use GitHub?

https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account

If you need more information on how to use ssh, you can generate a new ssh key using this GitHub documentation

https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent

Are you saying that you have not been able to use GitHub from your Mac for the past year?

**Comment by jcasman** [comment]
I've been working from the Dell Windows laptop

**Comment by AirTechWick** [comment]
This way might be easier when updating your own local branch to the most up to date main branch

https://stackoverflow.com/questions/3876977/update-git-branches-from-master

![image](https://user-images.githubusercontent.com/82003999/225708796-e407e254-d1c4-4066-b692-0eb6c4ad9cda.png)

**Reply by codetricity** [reply]
@AirTechWick  `git rebase master` looks good.  I've been trying to use it.  I like how the question has 998 up votes, there are 15 answers and the top answer has 946 upvotes.  We're not alone in our git journey.

BTW, I've noticed that there's a _Revert_ button on the commits.

![image](https://user-images.githubusercontent.com/5799860/225765333-ccc453b8-d8f0-4935-be91-a21409c67d15.png)

This document was also interesting.

https://github.com/git/git/blob/master/Documentation/howto/revert-a-faulty-merge.txt

**Reply by AirTechWick** [reply]
Looks like it should work

**Comment by AirTechWick** [comment]
I think what I was having trouble with is when I had "merge conflicts" and when I tried to resolve the merge conflicts on my branch I made mistakes doing it.

**Reply by AirTechWick** [reply]
On my branch it was "behind commits" because, the remote main had new updates, and it was also "ahead of commits' because I had worked on features without pushing I believe. I'll be checking out the [Github Desktop App ](https://desktop.github.com/) to see if resolving "merging conflicts" might be easier there for me.

**Reply by AirTechWick** [reply]
https://www.youtube.com/watch?v=JtIX3HJKwfo

From watching this video he just deletes the added extra  "<<<" that github/git generates

**Reply by codetricity** [reply]
Thank you for doing this research.  We're in a small team of three people (@jcasman , @codetricity , @AirTechWick ) with a shared goal and shared challenges.  On Monday, I suggest that you try a merge soon after you get in and that @jcasman sits by you and helps you get through the initial fetch, merge, change, and pull request cycle.  

Even after the pull request is successfully merged, I suggest we still keep practices and in fact think of a realistic scenario to introduce a merge conflict.  For example, @jcasman is working on a demo-react-native Mac install document.  He could try and make a small edit to 

https://github.com/theta360developers/theta-client-guide/blob/main/docs/react-native-install.md

To set up a placeholder for the Mac section.

One possible way to view this is option 1 or option 2

![image](https://user-images.githubusercontent.com/5799860/225964040-98958e11-3c10-4b13-a066-279d2191ed86.png)

BTW, I've watched videos from this guy before on the Processing language. I've also read his book:

https://natureofcode.com/book/

**Reply by AirTechWick** [reply]
That screenshot is very helpful when you separate the two options like that. Makes more sense to me now.

**Comment by AirTechWick** [comment]
May be useful if you want further protection of the main branch.
https://github.blog/2018-03-23-require-multiple-reviewers/
![image](https://user-images.githubusercontent.com/82003999/226403586-3edc1847-07f1-4f94-92b5-386ca205f22b.png)

**Reply by codetricity** [reply]
For our purposes, we should have just 1 reviewer now.  In the future, if there are multiple interns, we may have multiple reviewers of interns in a peer-based review structure.

At the moment, I suggest we require only 1 reviewer.

**Reply by jcasman** [reply]
Agree, only 1 reviewer should be required at this point.

**Comment by jcasman** [comment]
I had a problem with `git branch -a`  showing old information. I could see a branch called `craig-flutter-preview`  through GitHub (web) but running `git branch -a`  did not show it.

I ran `git remote update`  and `git branch -a`  now shows `craig-flutter-preview`

**Reply by codetricity** [reply]
theoretically, `git pull` does both `git remote update` and `git fetch`.  Did you still have the problem after you ran `git pull`?

https://www.janbasktraining.com/community/devops/what-is-the-difference-between-git-remote-update-git-fetch-and-git-pull

**Reply by jcasman** [reply]
It worked. This is super vague but sometimes it seems like `git pull` doesn't work and then I try it again a little later and it does. Anyway, it works.

**Reply by codetricity** [reply]
Can you consistently show that the following does not update the branch information?

```
git pull
git branch -a
```

**Comment by AirTechWick** [comment]
Seems like many people have this same problem
https://github.com/orgs/community/discussions/26693
![image](https://user-images.githubusercontent.com/82003999/229574332-a20a4457-e27a-4141-bb84-c2dc5088797f.png)

**Reply by jcasman** [reply]
I pressed the button and tried the pulldown menu. It "failed to cancel workflow" multiple times.

Re-running the Action (running `mkdocs gh-deploy` from the command line) worked fine.

Queued Action is still there, cannot be cancelled, cannot be deleted. I believe this would not stop us from just moving forward. Erik requesting GitHub support to delete stuck workflow.

**Reply by AirTechWick** [reply]
Yes, sent in a support ticket.

**Reply by codetricity** [reply]
This is a minor problem right?  You can just start the action again, right?

**Reply by jcasman** [reply]
Yes, it appears as if we can just run the action again.

**Reply by codetricity** [reply]
I've seen this before a number of times in the same situation.  I've been running `mkdocs gh-deploy` again to deploy it.

**Comment by AirTechWick** [comment]
Whenever updating my Local main branch to the up to date Remote main branch this is the command I use in-case I get the brown warning message.
`git pull origin main --rebase`
Works well for me.
![image](https://user-images.githubusercontent.com/82003999/230963278-ee920075-bfb3-4019-b1c7-ee2920d2d9a6.png)

**Reply by codetricity** [reply]
Thanks for the tip.  When you run these commands, are you on your own branch (example: erik-mkdocs-upgrade)?

**Comment by AirTechWick** [comment]
https://stackoverflow.com/questions/9537392/git-fetch-a-remote-branch
![image](https://github.com/theta360developers/theta-client-guide/assets/82003999/673c4762-bf75-405b-b4b1-23d3a88fe7ce)
`git switch` gets a remote branch to my local computer and tracks it automatically.
